### PR TITLE
Add support for spec middleware (form, example)

### PR DESCRIPTION
**Pull request by christoph-frick**

### CODE REVIEW

**Comment by christoph-frick** [comment]
This is a first attempt in supporting the features from the `spec` middleware.

One problem so far is the cursor jumping to the beginning of the file after `cs[fe]`.

So far i tested with:

```clojure
(ns vim-fireplace-spec.core
  (:require [clojure.spec.alpha :as s]))

(alias 'self *ns*)

(s/def ::id int?)
(s/def ::name string?)
(s/def ::user (s/keys :req [::id ::name]))

#_(
   ::id
   ::self/id
   :vim-fireplace-spec.core/id
   ::name
   ::self/name
   :vim-fireplace-spec.core/name
   ::user
   ::self/user
   :vim-fireplace-spec.core/user
   )

#_ (->> (s/registry) (keys) (map str))
```

**Comment by tpope** [comment]
I'm out of the loop, can you link to the documentation of what this is spec middleware is exactly?

The cursor jump is because functions in VimL return `0` by default, and `:exe 0` is the same as `:0` which jumps to the first line. `return ''` will fix it.

**Comment by christoph-frick** [comment]
Relevant links:

- https://github.com/clojure-emacs/cider-nrepl/blob/master/src/cider/nrepl/middleware/spec.clj
- https://github.com/clojure-emacs/orchard/blob/master/src/orchard/spec.clj

I have no clue how that looks in emacs.

Thanks for the hint about the jump.  I'll fix that.

**Comment by christoph-frick** [comment]
I have force-pushed the fix with the `return ""`.

File: plugin/fireplace.vim [hunk 1]
```diff
@@ -1666,6 +1666,42 @@ augroup fireplace_go_to_file
   autocmd FileType clojure call s:set_up_go_to_file()
 augroup END
 
+" Section: Spec
+" TODO: `spec-list`
+
+function! fireplace#spec(op, symbol) abort
+  if fireplace#op_available(a:op)
+    let unaliased_symbol = fireplace#eval(a:symbol)
```
Comment (tpope) [inline plugin/fireplace.vim]: Eval is a really crude hammer to swing here. Try something like this:
```vim
  if a:symbol =~# '^::.\+/'
    let symbol = ':' . fireplace#resolve_alias(matchstr(a:symbol, '^::\zs[^/]\+')) . matchstr(a:symbol, '/.*')
  elseif a:symbol =~# '^::'
    let symbol = ':' . fireplace#ns() . '/' . strpart(a:symbol, 2)
  else
    let symbol = a:symbol
  endif
```

It won't handle `(alias 'self *ns*)`, but I'll worry about that when I see it in production code.

File: plugin/fireplace.vim [hunk 2]
```diff
@@ -1666,6 +1666,42 @@ augroup fireplace_go_to_file
...
+    let response = fireplace#message({'op': a:op, 'spec-name': unaliased_symbol})[0]
+    if !empty(get(response, a:op))
+      echo get(response, a:op)
```
Comment (tpope) [inline plugin/fireplace.vim]: For `spec-form`, this is just dumping a raw Vim data structure. Surely we can do better?

File: plugin/fireplace.vim [hunk 3]
```diff
@@ -1666,6 +1666,42 @@ augroup fireplace_go_to_file
...
+  call fireplace#spec("spec-example", a:symbol)
+  return ''
+endfunction
+
+nnoremap <silent> <Plug>FireplaceSpecForm    :<C-U>exe <SID>SpecForm(expand('<cword>'))<CR>
+nnoremap <silent> <Plug>FireplaceSpecExample :<C-U>exe <SID>SpecExample(expand('<cword>'))<CR>
+
+function! s:set_up_fireplace_spec() abort
+  call s:map('n', 'csf', '<Plug>FireplaceSpecForm')
+  call s:map('n', 'cse', '<Plug>FireplaceSpecExample')
```
Comment (tpope) [inline plugin/fireplace.vim]: These maps conflict with surround.vim and thus are disqualified in my book. For `spec-form`, it might make sense to overload `K`.

I think more fundamental than maps is providing commands like `:SpecName`.

File: plugin/fireplace.vim (hunk 2 above)
Comment (christoph-frick) [reply plugin/fireplace.vim]: The "SpecExample" result is sent through the pretty printer in orchard/cider. I'll separate the two functions and see how i can make the SpecForm result more clojury

File: plugin/fireplace.vim (hunk 3 above)
Comment (christoph-frick) [reply plugin/fireplace.vim]: With `K` you mean adding a third branch into `s:K` and dispatch to SpecForm if `word` starts with `:`?

For now i'll remove the keyboard shortcuts and add the commands.

File: plugin/fireplace.vim (hunk 1 above)
Comment (christoph-frick) [reply plugin/fireplace.vim]: Just for the records: `(alias 'self *ns*)` is just a hack to have a single file but be able to test `::nsalias/kw`, which is became way more likely since spec.  More realistic example

```clojure
(ns com.example.game.board (:require [com.example.game.player :as player] ...))
(s/def ::board (s/keys :req [::player/players]))
```

File: plugin/fireplace.vim (hunk 2 above)
Comment (christoph-frick) [reply plugin/fireplace.vim]: If this was a rethorical question, i need some hints how to deal with the output

File: plugin/fireplace.vim (hunk 3 above)
Comment (tpope) [reply plugin/fireplace.vim]: Right. Currently pressing `K` on a keyword is worthless so why not?

File: plugin/fireplace.vim (hunk 2 above)
Comment (tpope) [reply plugin/fireplace.vim]: I assume what's needed is some sort of recursive function that turns `['a', ['b', 'c']]' into `(a (b c))`.

File: plugin/fireplace.vim (hunk 1 above)
Comment (tpope) [reply plugin/fireplace.vim]: Exactly what I figured. I think we're good.

File: plugin/fireplace.vim (hunk 2 above)
Comment (tpope) [reply plugin/fireplace.vim]: Something like this perhaps:

```vim
function! s:pr(obj) abort
  if type(a:obj) == v:t_string
    return a:obj
  elseif type(a:obj) == v:t_list
    return '(' . join(map(copy(a:obj), 's:pr(v:val)'), ' ') . ')'
  elseif type(a:obj) == v:t_dict
    return '{' . join(map(keys(a:obj), 's:pr(v:val) . " " . s:pr(a:obj[v:val])'), ', ') . '}'
  else
    return string(a:obj)
  endif
endfunction
```

It does the right thing for plain strings too so no need to special case.

File: plugin/fireplace.vim [hunk 4]
```diff
@@ -1666,6 +1666,62 @@ augroup fireplace_go_to_file
...
+" TODO: `spec-list`
+
+function! fireplace#symbol_or_fallback(...) abort
+  if empty(a:1)
+    return expand('<cword>')
+  endif
+  return a:1
+endfunction
+
+function! fireplace#fully_qualified_symbol(symbol) abort
```
Comment (tpope) [inline plugin/fireplace.vim]: `qualify_keyword` seems like a more accurate name.

File: plugin/fireplace.vim [hunk 5]
```diff
@@ -1666,6 +1666,62 @@ augroup fireplace_go_to_file
...
+    let response = fireplace#message({'op': op, 'spec-name': symbol})[0]
+    if !empty(get(response, op))
+      echo get(response, op)
+    endif
+  endif
+  return ''
+endfunction
+
+function! s:set_up_fireplace_spec() abort
+  command! -buffer -bar -nargs=? -complete=customlist,fireplace#eval_complete SpecForm    :exe s:SpecForm(<q-args>)
```
Comment (tpope) [inline plugin/fireplace.vim]: Let's leave this as `-nargs=1` for now. I won't rule out the `<cword>` default but we should do it everywhere it makes sense if so.

**Comment by christoph-frick** [comment]
I have addressed all your points

File: plugin/fireplace.vim [hunk 6]
```diff
@@ -1669,41 +1681,33 @@ augroup END
 " Section: Spec
 " TODO: `spec-list`
```
Comment (tpope) [inline plugin/fireplace.vim]: What's the story here? I would imagine that switching from maps to commands would make this one easy to implement?

Comment (christoph-frick) [reply plugin/fireplace.vim]: Removed that comment.  It's basically `(spec/registry)` and I assume, that in emacs/cider one opens that list and then navgiates to form/example from there.  AFAIK there currently is nothing similar in fireplace to add this to.

File: doc/fireplace.txt [hunk 7]
```diff
@@ -57,11 +57,19 @@ DOCUMENTATION                                   *fireplace-documentation*
 :Javadoc {class}        Open the java docs for the given class in a browser.
 
                                                 *fireplace-K*
-K                       Look up docs for symbol under cursor.
+K                       Look up (java) doc/spec-form for symbol/keyword under
+                        cursor.
```
Comment (tpope) [inline doc/fireplace.txt]: I found this a tad hard to parse. How about `Look up the doc, javadoc, or spec-form for the identifier under the cursor.`?

Comment (christoph-frick) [reply doc/fireplace.txt]: Sure

File: doc/fireplace.txt [hunk 8]
```diff
@@ -57,11 +57,19 @@ DOCUMENTATION                                   *fireplace-documentation*
...
                                                 *fireplace-:FindDoc*
 :FindDoc {arg}          Wrapper around (clojure.repl/find-doc ...).
 
+                                                *fireplace-:SpecForm*
+:SpecForm [kw]          Show form for spec at the cursor or given argument
+
+                                                *fireplace-:SpecExample*
+:SpecExample [kw]       Generate example for spec at the cursor or given
+                        argument
+
```
Comment (tpope) [inline doc/fireplace.txt]: `:SpecForm {keyword}`, and the wording needs tweaking but I'm not sure to what. Do these roughly correspond to `clojure.spec` functions? If so, it might make sense to reference that like we do for `:FindDoc`. I worry that just saying "spec" is confusing to those not familiar (it was for me).

Comment (christoph-frick) [reply doc/fireplace.txt]: We could use some codified version like `(spec/form)` - they are at least all named the same for now in all three versions of spec, that are in the wild.

Comment (christoph-frick) [reply doc/fireplace.txt]: I used `spec/form` and `gen/generate` - the later is the alias used in the main docs https://clojure.org/guides/spec#_sampling_generators; the first is a bit more clear than just `s`

**Comment by tpope** [comment]
I'm afraid I saved the hardest part for last: error handling. We need to handle the case the op isn't there. This is the first time we don't have a fallback for if CIDER isn't installed, which is fine, but I need to think a bit on how to proceed.

We also need error handling for if the op is there but fails. What are the failure cases, and how can we handle them?

File: doc/fireplace.txt (hunk 8 above)
Comment (tpope) [reply doc/fireplace.txt]: Can we go all in and use the full namespaces? That should make it easier to look up.

Comment (christoph-frick) [reply doc/fireplace.txt]: The problem with the namespace would be to name all two currently (and soon three) supported by orchard/cider (spec "0", spec alpha, spec alpha 2) - which is a problem on it's on (see answer to "error handling" comment below)

**Comment by christoph-frick** [comment]
I'll list all i thought about so far:
- Op is not there -> handled already, but we could handle this ourself (see below); use fallback (see below) or notify user?
- a spec kw not known (currently we fail silently without giving the user feedback)
- a spec kw not loaded yet (same as "not known", but might be confusing for the user) - I think we should keep it that way, since we can not know where the def is and load it for the user; there is autocompletion in the prompt which makes it more obvious
- there is an error somewhere down the road and something throws -> think this is handled already: random errors bubble up
- generating an example ends in an "infinite loop" - if the spec is complex or recursive and the user does not take care; the example gets huge, eats all the CPU and RAM. I'd expect CTRL-C works here and propagates well into cider (?)

As for the fallback: no cider or old cider: we can do the same as [orchard](https://github.com/clojure-emacs/orchard/blob/master/src/orchard/spec.clj) does.  We could Eval some "try what spec is there" with the given keyword.

The only problem I see with this (and it's one with orchard too): the search for the "right" spec lib just favours the order of existence (try alpha1, then spec0 - soon to be started with alpha2).  Right now this is not a huge problem, because most likely folks will use alpha1.  But once alpha2 hits, the code bases are spec1 and spec2 "is there" and maybe incompatible and in the worst case scenario messes things up if loaded in parallel.  So that could mean, a switch per project is needed to tell the tooling, which spec version is in use or have the tooling do more work in finding the right one.

**Comment by tpope** [comment]
The current "op is not there" behavior is silent failure too. We need separate errors for both not connected and connected but no op. This is the thing I will think on. (If you want to provide an eval callback, great, but I'm inclined to think it's not worth bothering with.)

The only other case I care about is "not known". The others seem fine as is.

File: plugin/fireplace.vim (hunk 3 above)
Comment (christoph-frick) [reply plugin/fireplace.vim]: > Right. Currently pressing `K` on a keyword is worthless so why not?

For the records: this is actually the behaviour of `clojure.repl/doc` (if arg is a keyword, show `spec/describe`

File: plugin/fireplace.vim [hunk 9]
```diff
@@ -1666,6 +1678,53 @@ augroup fireplace_go_to_file
...
+    let kw = a:kw
+  endif
+  return kw
+endfunction
+
+function! s:SpecForm(kw) abort
+  let op = "spec-form"
+  if fireplace#op_available(op)
+    let symbol = fireplace#qualify_keyword(a:kw)
+    let response = fireplace#message({'op': op, 'spec-name': symbol})[0]
```
Comment (tpope) [inline plugin/fireplace.vim]: Use the new `fireplace#message({...}, v:t_dict)` API here to get back a single dictionary, rather than relying on the response to be a single message.

File: plugin/fireplace.vim [hunk 10]
```diff
@@ -1666,6 +1678,53 @@ augroup fireplace_go_to_file
...
+    if !empty(get(response, op))
+      echo s:pr(get(response, op))
```
Comment (tpope) [inline plugin/fireplace.vim]: An unknown op error includes the original op, so this should be adequate for error handling:

```viml
elseif has_key(response, 'op')
  return 'echoerr ' . string('Fireplace: no nREPL op available for ' . op)
````

Use this same `echoerr` pattern for any other error handling.

File: plugin/fireplace.vim [hunk 11]
```diff
@@ -1666,6 +1678,53 @@ augroup fireplace_go_to_file
...
+    endif
+  endif
```
Comment (tpope) [inline plugin/fireplace.vim]: Drop the conditional and wrap everything atry block to handle the case when we're not connected:

```viml
try
  ...
catch /^Fireplace:/
  return 'echoerr ' . v:exception
entry
```

**Comment by christoph-frick** [comment]
I have addressed above three points and tested them (still works, unknown op gives an proper error, no repl gives an proper error).

**Comment by tpope** [comment]
Thanks for following through!

**Comment by christoph-frick** [comment]
Cheers
